File: 34_Find_First_And_Last_Element_in_Sorted_Array/Solution.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        

        int[] ans = {-1,-1};
        int start = search(nums,target,true);
        int end = search(nums, target, false);
        ans[0] = start;
        ans[1] = end; 
        return ans;

    }
    int search(int[] nums, int target, boolean search_for_left){
        int left = 0;
        int right = nums.length-1;
        int index = -1;
        while (left <= right){
            int mid = left+(right-left)/2;
            if (nums[mid] > target) {
                right = mid-1;
            } else if (nums[mid]<target){
                left = mid+1;
            } else {
                index = mid;
                if (search_for_left){
                    right = mid-1;
                } else {
                    left = mid+1;
                }
            }
        }
        return index;
    }
}

File: 34_Find_First_And_Last_Element_in_Sorted_Array/Solution.cpp
class Solution {
public:
    int findLeftMost(vector<int>& nums, int target, int n){
        int l = 0, r = n-1;
        int l_idx = -1;
        while(l<=r){
            int mid = l+(r-l)/2;
            if(nums[mid] == target){
                l_idx = mid;
                r = mid-1;
            } else if(nums[mid] < target){
                l = mid+1;
            } else {
                r = mid-1;
            }
        }
        return l_idx;
    }
    int findRightMost(vector<int>& nums, int target, int n){
        int l = 0, r = n-1;
        int r_idx = -1;
        while(l<=r){
            int mid = l+(r-l)/2;
            if(nums[mid] == target){
                r_idx = mid;
                l = mid+1;
            } else if(nums[mid] < target){
                l = mid+1;
            } else {
                r = mid-1;
            }
        }
        return r_idx;
    }
    vector<int> searchRange(vector<int>& nums, int target) {
        int n = nums.size();
        int left_most = findLeftMost(nums,target,n);
        int right_most = findRightMost(nums,target,n);

        return {left_most,right_most};
    }
};